src/zenml/integrations/modal/step_operators/modal_step_operator.py

```diff
@@ -0,0 +1,215 @@
+                "it is correctly configured."
+            )
+
+        if docker_creds := stack.container_registry.credentials:
+            docker_username, docker_password = docker_creds
+        else:
+            raise ValueError(
+                "No Docker credentials found for the container registry."
+            )
+
+        # get the pydantic version in local environment
+        # use it to install the correct version of pydantic in the modal app
+        # since it overwrites with 1.x on top of our image
+        pydantic_version = importlib.metadata.version("pydantic")
+        major, minor, *_ = pydantic_version.split(".")
+
+        # Construct the decorator arguments based on the settings
+        decorator_args = []
+        if settings.region is not None:
+            decorator_args.append(f"region='{settings.region}'")
+        if settings.cloud is not None:
+            decorator_args.append(f"cloud='{settings.cloud}'")
+
+        # if resource settings are set, add them to the decorator args
+        resource_settings = info.config.resource_settings
+        if settings.gpu is not None:
+            gpu_str = settings.gpu
+            if resource_settings.gpu_count is not None:
+                gpu_str += f":{resource_settings.gpu_count}"
+            decorator_args.append(f"gpu='{gpu_str}'")
+        if resource_settings.cpu_count is not None:
+            decorator_args.append(f"cpu={resource_settings.cpu_count}")
+        if resource_settings.memory is not None:
+            decorator_args.append(
+                f"memory={resource_settings.get_memory(ByteUnit.MB)}"
+            )
+
+        decorator_args_str = ", ".join(decorator_args)
+
+        with tempfile.NamedTemporaryFile(
+            mode="w", delete=False, suffix=".py"
+        ) as tmp:
+            modal_code = f"""import modal
+
+my_secret = modal.Secret.from_dict({{
+    'REGISTRY_USERNAME': '{docker_username}',
+    'REGISTRY_PASSWORD': '{docker_password}',
+}})
+
+zenml_image = modal.Image.from_registry(tag='{image_name}', secret=my_secret).env({environment}).pip_install("pydantic~={major}.{minor}").run_commands('{" ".join(entrypoint_command)}')
+
+app = modal.App('{info.run_name}')
+
+@app.function(image=zenml_image, {decorator_args_str})
+def run_step():
+    print("Executing {info.run_name} step...")
+
+if __name__ == "__main__":
+    run_step()
+"""
+            tmp.write(modal_code)
+            tmp_filename = tmp.name
+
+        # Run the Modal app using subprocess
+        try:
+            subprocess.run(
+                ["modal", "run", tmp_filename],
+                check=True,
+                capture_output=True,
+                text=True,
+            )
+            logger.info("Modal app completed successfully.")
+        except subprocess.CalledProcessError as e:
+            logger.error(f"Modal app exited with code {e.returncode}.")
+            logger.error(e.stdout)
+            raise
```

Comment:
Maybe we should have a custom exception here that explains that our step operator failed?

```python
raise RuntimeError(f"Model step operator failed: {str(e)}")
```
or something like that?

Reply:
Yeah makes sense